### PR TITLE
VarInt/VarLong Encode and Decode Optimizations

### PR DESCRIPTION
**Pull request by guac42**
Aims to improve upon VarInt/VarLong Encode and Decode procedures (#208).
# Baselines:
## VarInt:
VarInt::decode
VarInt::encode
## VarLong
VarLong::encode
VarLong::decode

### CODE REVIEW

**Comment by guac42** [comment]
image
Performance improvement of VarInt::encode on my machine. The original version was running at about 39ns.

**Comment by guac42** [comment]
I tried a similar encode approach on VarLong and it resulted in some regression probably because it requires the use of many u128 bitwise ops.

**Comment by Moulberry** [comment]
The code is not mine, I must have removed some comments during refactoring.
It comes from https://github.com/as-com/varint-simd/, specifically 
https://github.com/as-com/varint-simd/blob/0f468783da8e181929b01b9c6e9f741c1fe09825/src/encode/mod.rs#L71

**Comment by guac42** [comment]
image
Performance improvement of VarLong::encode on my machine. The original version was running at about 40ns.

**Comment by guac42** [comment]
I tested branchless approaches to VarInt::decode and VarLong::decode and saw either no improvement or regression.

File: crates/valence_protocol/src/var_int.rs [hunk 1]
```diff
@@ -48,16 +48,31 @@ pub enum VarIntDecodeError {
...
-        let mut val = self.0 as u32;
-        loop {
-            if val & 0b11111111111111111111111110000000 == 0 {
-                w.write_u8(val as u8)?;
-                return Ok(());
-            }
-            w.write_u8(val as u8 & 0b01111111 | 0b10000000)?;
-            val >>= 7;
-        }
+        let x = unsafe { std::mem::transmute::<i32, u32>(self.0) } as u64;
```
Comment (rj00a) [inline crates/valence_protocol/src/var_int.rs:54]: If I'm not mistaken, this should be equivalent to `self.0 as u64`.

File: crates/valence_protocol/src/var_int.rs [hunk 2]
```diff
@@ -48,16 +48,31 @@ pub enum VarIntDecodeError {
...
+        let bytes_needed = 8 - unused_bytes;
+
+        // set all but the last MSBs
+        let msbs = 0x8080808080808080;
+        let msbmask = 0xffffffffffffffff >> (((8 - bytes_needed + 1) << 3) - 1);
+
+        let merged = stage1 | (msbs & msbmask);
+        let bytes = merged.to_le_bytes();
+
+        w.write_all(&bytes[..bytes_needed as usize])?;
```
Comment (rj00a) [inline crates/valence_protocol/src/var_int.rs:73]: There might be bounds checks happening here in the slicing that aren't optimized away. Try `get_unchecked`?

File: crates/valence_protocol/src/var_long.rs [hunk 3]
```diff
@@ -26,7 +30,59 @@ impl VarLong {
...
+        // Shift it down 1 byte so the last MSB is the only one set, and make sure only
+        // the MSB is set
+        let shift = unsafe { _mm_bsrli_si128(mask, 1) };
+        let msbmask = unsafe { _mm_and_si128(shift, _mm_set1_epi8(128u8 as i8)) };
+
+        // Merge the MSB bits into the vector
+        let merged = unsafe { _mm_or_si128(stage1, msbmask) };
+        let bytes = unsafe { std::mem::transmute::<__m128i, [u8; 16]>(merged) };
+
+        w.write_all(&bytes[..bytes_needed as usize])?;
```
Comment (rj00a) [inline crates/valence_protocol/src/var_long.rs:74]: Same bounds check issue.

**Review by rj00a — COMMENTED** [review]
- Can you include the old approaches in the benchmarks for comparison?
- Have you tried the technique described in "how fast can you encode a varint"? The author reported a 30% perf improvement in their environment.

File: crates/valence_protocol/src/var_long.rs [hunk 4]
```diff
@@ -26,7 +30,59 @@ impl VarLong {
 }
 
 impl Encode for VarLong {
+    // Adapted from VarInt-Simd encode
+    // https://github.com/as-com/varint-simd/blob/0f468783da8e181929b01b9c6e9f741c1fe09825/src/encode/mod.rs#L71
+    #[cfg(all(
+        any(target_arch = "x86", target_arch = "x86_64"),
+        not(target_os = "macos")
+    ))]
     fn encode(&self, mut w: impl Write) -> Result<()> {
```
Comment (rj00a) [inline crates/valence_protocol/src/var_long.rs:39]: Looks like we're missing a fallback implementation for non-x86 or non-x86_64 platforms. If you decide to use this approach, please make a generic implementation that works on all platforms.

Comment (guac42) [reply crates/valence_protocol/src/var_long.rs:39]: Would this work for the fallback implementation?
```
#[cfg(any(
    not(any(target_arch = "x86", target_arch = "x86_64")),
    target_os = "macos"
))]
```

File: crates/valence_protocol/src/var_int.rs (hunk 1 above)
Comment (guac42) [reply crates/valence_protocol/src/var_int.rs:54]: Yea that's correct. I should have noticed that LOL.

File: crates/valence_protocol/src/var_long.rs (hunk 4 above)
Comment (rj00a) [reply crates/valence_protocol/src/var_long.rs:39]: Oh wait I didn't read the rest of the code. You're fine.

File: crates/valence_protocol/src/var_long.rs [hunk 5]
```diff
@@ -1,7 +1,11 @@
+#[cfg(target_arch = "x86")]
+use std::arch::x86::*;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::*;
```
Comment (rj00a) [inline crates/valence_protocol/src/var_long.rs:4]: You can scope these imports to the body of `encode` so you don't have to use `cfg` here.

**Comment by guac42** [comment]
> Have you tried the technique described in "how fast can you encode a varint"?

Yes, I think the performance the author gained was mostly due to java inefficiencies.
Here is the rust implementation's performance:
image

**Comment by rj00a** [comment]
> Here is the rust implementation's performance:

Can you include this code (and the old VarInt VarLong functions) in the criterion benchmarks?

Don't forget to mark this PR as "ready for review" if it is finished.

**Comment by guac42** [comment]
> Can you include this code in the criterion benchmarks?

I have since written a more optimized version of what I had previously benched. It's still based on [how fast can you encode a varint?](https://steinborn.me/posts/performance/how-fast-can-you-write-a-varint/).
### Here's the performance:
image
### Here's the code:
```rust
fn encode(&self, mut w: impl Write) -> anyhow::Result<()> {
    if (self.0 as u32 & (0xffffffff << 7)) == 0 {
        w.write_u8(self.0 as u8)?;
    } else if (self.0 as u32 & (0xffffffff << 14)) == 0 {
        let val = (self.0 as u32 & 0x7F | 0x80) << 8 | (self.0 as u32 >> 7);
        w.write_u16::<BigEndian>(val as u16)?;
    } else if (self.0 as u32 & (0xffffffff << 21)) == 0 {
        let val = (self.0 as u32 & 0x7F | 0x80) << 16 | ((self.0 as u32 >> 7) & 0x7f | 0x80) << 8 | (self.0 as u32 >> 14);
        w.write_u24::<BigEndian>(val)?;
    } else if (self.0 as u32 & (0xffffffff << 28)) == 0 {
        let val = (self.0 as u32 & 0x7F | 0x80) << 24 | (((self.0 as u32 >> 7) & 0x7f | 0x80) << 16)
        | ((self.0 as u32 >> 14) & 0x7F | 0x80) << 8 | (self.0 as u32 >> 21);
        w.write_u32::<BigEndian>(val)?;
    } else {
        let val = ((self.0 as u64) & 0x7f | 0x80) | ((self.0 as u64 >> 7) & 0x7f | 0x80) << 8
            | ((self.0 as u64 >> 14) & 0x7f | 0x80) << 16 | ((self.0 as u64 >> 21) & 0x7f | 0x80) << 24
            | ((self.0 as u64 >> 28) & 0x7f) << 32;
        w.write_all(unsafe { val.to_le_bytes().get_unchecked(..5usize) })?;
    }
}
```

**Comment by guac42** [comment]
I even tried going as far as to bypass rusts arithmetic checks using ```u32::wrapping_add``` etc. And didn't achieve any real performance gains. :/ 
But 5% is still a fair amount, I guess.

**Comment by rj00a** [comment]
Thanks for taking the time to check. Btw arithmetic checks only happen in debug mode so `wrapping_add` is equivalent to `+` in release mode.

**Pull request merged**